File: iosApp/iosApp/presentationLayer/pages/settings/viewModels/SettingsViewModel.swift
import Foundation
import shared
import KMPNativeCoroutinesAsync

class SettingsViewModel: SettingsViewModelProtocol {

    @Published var viewState: SettingsViewState = .initial(
        accountShortCodeForShareTitle: "carregando...",
        accountsSharedWithMeTitle: "Acessar conta compartilhada",
        accountsSharedWithMeSubtitle: "Obs: atualmente limitada apenas a 1 conta"
    )
    @Published var myAccount: AccountModel?

    var goBack: () -> Void = {}

    private let getAccountUseCase: GetAccountUseCase
    private let getLocalAccountUuidUseCase: GetLocalAccountUuidUseCase
    private let setSharedAccountByCodeUseCase: SetSharedAccountByCodeUseCase
    private let crashlytics: IFirebaseCrashlytics

    init(
        getAccountUseCase: GetAccountUseCase,
        getLocalAccountUuidUseCase: GetLocalAccountUuidUseCase,
        setSharedAccountByCodeUseCase: SetSharedAccountByCodeUseCase,
        crashlytics: IFirebaseCrashlytics
    ) {
        self.getAccountUseCase = getAccountUseCase
        self.getLocalAccountUuidUseCase = getLocalAccountUuidUseCase
        self.setSharedAccountByCodeUseCase = setSharedAccountByCodeUseCase
        self.crashlytics = crashlytics
    }

    @MainActor
    func getAccount() async {

        viewState = .loading

        do {
            let accountUuid = try await asyncFunction(for: getLocalAccountUuidUseCase.execute())
            myAccount = try await asyncFunction(for: getAccountUseCase.execute(accountUuid: accountUuid))
        } catch {
            showError(e: error)
            return
        }

        let accountShortCodeForShareTitle = self.myAccount?.accountShortCodeForShare ?? "carregando..."
        var accountsSharedWithMeTitle = "Acessar conta compartilhada"
        var accountsSharedWithMeSubtitle = "Obs: atualmente limitada apenas a 1 conta"

        if let sharedAccount = self.myAccount?.accountsSharedWithMe.first {
            accountsSharedWithMeTitle = "Acessando a conta:"
            accountsSharedWithMeSubtitle = sharedAccount
        }

        viewState = .success(
            accountShortCodeForShareTitle: accountShortCodeForShareTitle,
            accountsSharedWithMeTitle: accountsSharedWithMeTitle,
            accountsSharedWithMeSubtitle: accountsSharedWithMeSubtitle
        )
    }

    @MainActor
    func accessSharedAccountWithCode(code: String) async {

        viewState = .loading

        do {
            let accountUuid = try await asyncFunction(for: getLocalAccountUuidUseCase.execute())
            _ = try await asyncFunction(for: setSharedAccountByCodeUseCase.execute(accountUuid: accountUuid, code: code))
            await getAccount()
        } catch {
            showError(e: error)
            return
        }
    }

    func shareMyCode() {
        if let _ = myAccount?.accountShortCodeForShare {
            //TODO: Implement sharing functionality
        }
    }

    func openAppHomePage() {
        // TODO: mplement opening app home page functionality
    }

    func showError(e: Error) {
        crashlytics.record(error: e)
        viewState = .error(message: e.localizedDescription, retryAction: {
            Task {
                await self.getAccount()
            }
        })
    }
}

File: androidApp/src/main/java/dev/haroldjose/familysharedlist/android/presentationLayer/pages/settings/viewmodels/SettingsViewModel.kt
package dev.haroldjose.familysharedlist.android.presentationLayer.pages.settings.viewmodels

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.haroldjose.familysharedlist.Logger
import dev.haroldjose.familysharedlist.android.app.MainApplication
import dev.haroldjose.familysharedlist.domainLayer.models.AccountModel
import dev.haroldjose.familysharedlist.domainLayer.usecases.account.GetAccountUseCase
import dev.haroldjose.familysharedlist.domainLayer.usecases.account.GetLocalAccountUuidUseCase
import dev.haroldjose.familysharedlist.domainLayer.usecases.account.SetSharedAccountByCodeUseCase
import dev.haroldjose.familysharedlist.getPlatform
import dev.haroldjose.familysharedlist.services.firebase.IFirebaseCrashlytics
import kotlinx.coroutines.launch

class SettingsViewModel(
    private val getAccountUseCase: GetAccountUseCase,
    private val getLocalAccountUuidUseCase: GetLocalAccountUuidUseCase,
    private val setSharedAccountByCodeUseCase: SetSharedAccountByCodeUseCase,
    private val crashlytics: IFirebaseCrashlytics,
): ViewModel(), ISettingsViewModel {

    private val constAccountsSharedWithMeTitle = "Acessar conta compartilhada"
    private val constAccountsSharedWithMeSubtitle = "Obs: atualmente limitada apenas a 1 conta"
    private val constLoadingMessage = "carregando..."

    override var viewState: SettingsViewState by mutableStateOf(
        SettingsViewState.Initial(
            accountShortCodeForShareTitle = constLoadingMessage,
            accountsSharedWithMeTitle = constAccountsSharedWithMeTitle,
            accountsSharedWithMeSubtitle = constAccountsSharedWithMeSubtitle
        )
    )
    override var myAccount: AccountModel? = null

    override var goBack: () -> Unit = {}

    override suspend fun getAccount() {
        viewState = SettingsViewState.Loading
        try {
            val accountUuid = getLocalAccountUuidUseCase.execute()
            myAccount = getAccountUseCase.execute(accountUuid = accountUuid)

            var accountShortCodeForShareTitle = myAccount?.accountShortCodeForShare ?: constLoadingMessage
            var accountsSharedWithMeTitle = constAccountsSharedWithMeTitle
            var accountsSharedWithMeSubtitle = constAccountsSharedWithMeSubtitle
            myAccount?.accountsSharedWithMe?.let {
                if (it.isNotEmpty()) {
                    accountsSharedWithMeTitle = "Acessando a conta:"
                    accountsSharedWithMeSubtitle = it.first()
                }
            }
            viewState = SettingsViewState.Success(
                accountShortCodeForShareTitle = accountShortCodeForShareTitle,
                accountsSharedWithMeTitle = accountsSharedWithMeTitle,
                accountsSharedWithMeSubtitle = accountsSharedWithMeSubtitle
            )
        } catch (e: Throwable) {
            showError(e)
            return
        }
    }

    private fun showError(e: Throwable) {
        e.message?.let { Logger.d("showError", it) }
        crashlytics.record(e)
        viewState = SettingsViewState.Error(
            message = "Erro ao acessar a conta",
            retryAction = {
                this.viewModelScope.launch { getAccount() }
            }
        )
    }

    override suspend fun accessSharedAccountWithCode(code: String) {
        viewState = SettingsViewState.Loading
        try {
            val accountUuid = getLocalAccountUuidUseCase.execute()
            setSharedAccountByCodeUseCase
                .execute(
                    accountUuid = accountUuid,
                    code = code
                )
            getAccount()
        } catch (e: Throwable) {
            showError(e)
            return
        }
    }

    override suspend fun shareMyCode() {
        myAccount?.accountShortCodeForShare?.let {
            getPlatform().openShareOptionsWithText(text = it)
        }
    }

    override suspend fun openAppHomePage() {
        getPlatform().openUrlOnDefaultBrowser(url = "https://github.com/haroldjose30/FamilySharedList")
    }
    
    override fun getVersion(): String {
        try {
            MainApplication.applicationContext().let {
                val pInfo: PackageInfo = it.packageManager. getPackageInfo(it.getPackageName(), 0)
                val versionName = pInfo.versionName
                val longVersionCode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    pInfo.longVersionCode
                } else {
                    "0p"
                }
                return "Alpha $versionName (b$longVersionCode)"
            }
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            return ""
        }
    }
}